client/lib/create-selector/index.js

```diff
@@ -88,28 +86,25 @@ export default function createSelector(
 	getDependants = DEFAULT_GET_DEPENDANTS,
 	getCacheKey = DEFAULT_GET_CACHE_KEY
 ) {
-	const memoizedSelector = memoize( selector, getCacheKey );
-	let lastDependants;
+	const dependentsPerKey = new Map();
+	const memo = new Map();
 
 	if ( Array.isArray( getDependants ) ) {
 		getDependants = makeSelectorFromArray( getDependants );
 	}
 
-	return Object.assign(
-		function( state, ...args ) {
-			let currentDependants = getDependants( state, ...args );
-			if ( ! Array.isArray( currentDependants ) ) {
-				currentDependants = [ currentDependants ];
-			}
-
-			if ( lastDependants && ! shallowEqual( currentDependants, lastDependants ) ) {
-				memoizedSelector.cache.clear();
-			}
+	const memoizedSelector = function( state, ...args ) {
+		const cacheKey = getCacheKey( state, ...args );
+		const currentDependants = [].concat( getDependants( state, ...args ) );
```

Comment:
I think you could use [Lodash's `_.castArray`](https://lodash.com/docs/4.17.4#castArray) here to avoid needlessly allocating/concatenating an array if the result of `getDependants` already is one?

Reply:
I tested out the performance difference on [jsperf](https://jsperf.com/castarray-vs-concat).
You were right :).  castArray is much faster.  

Turns out writing it by hand is _even_ faster:
```js
const castArray = arg => Array.isArray( arg ) ? arg : [ arg ];
```

My guess would be that `Array.isArray` is faster than `_.isArray` 🤷‍♂️ .